doc/index.md

```diff
@@ -1,100 +1,178 @@
+
+- Errors are returned as `nil, error, errno` unless noted otherwise.
+
+- Some HTTP 2 operations return/throw special [http 2 error objects](#http.h2_error).
+
+
+## Terminology
+
+Much lua-http terminology is borrowed from HTTP 2.
+
+[Connection"](#connection) - An abstraction over an underlying TCP/IP socket. Lua-http has two connection types: one for HTTP 1, one for HTTP 2.
+
+[Stream"](#stream) - A request/response on a connection object. Lua-http has two stream types: one for [HTTP 1 streams](#http.h1_stream), and one for [HTTP 2 streams](#http.h2_stream). They common interfaces are descrined in [*stream*](#stream) and [*http.stream_common*](#http.stream_common).
 
 
 # Interfaces
 
+
 ## connection
 
 All connection types expose the fields:
 
+
 ### `connection.type` <!-- --> {#connection.type}
 
-Either `"client"` or `"server"`
+The mode of use for the connection object. Valid values are:
+- `"client"` - Connects to a remote uri
+- `"server"` - Listens for connection on a local uri
 
 
 ### `connection.version` <!-- --> {#connection.version}
 
-The HTTP version as a number
+Returns the HTTP version as a number. 
 
 
 ### `connection:connect(timeout)` <!-- --> {#connection:connect}
 
+Blocks (polls?) until the connection is complete or timeout is exceeded. Connecting may include DNS lookups, TLS negotiation and HTTP2 settings exchange. Returns true on success. On failure returns nil and an error message.
+
 
 ### `connection:checktls()` <!-- --> {#connection:checktls}
 
+Checks the socket for a valid Transport Layer Security connection. Return true if the connection is secured. Returns nil and an error message if there is no valid TLS encryption.
+
 
 ### `connection:localname()` <!-- --> {#connection:localname}
 
+Provides the uri for the local socket. The return format is: address family, IP address, port. If the connection object is not connected, returns AF_UNSPEC (0). If the error is other than a connection error, returns nil and the state name.
+
+    --example: print(connection:localname())
+    > 2 192.168.1.5 8000
+
 
 ### `connection:peername()` <!-- --> {#connection:peername}
 
+Provides the URI for the remote socket *peer* (as in the next hop).  Notably: If the client is using a proxy, the :peername() will be the proxy, not the remote server connection.The return format is: ip address, port. If not connected, returns 0. If the error is other than a connection error, returns nil and the statename.
+
 
 ### `connection:shutdown()` <!-- --> {#connection:shutdown}
 
+Closes all streams and calls shutdown on the socket. The connection cannot be re-opened. 
+
 
 ### `connection:close()` <!-- --> {#connection:close}
 
+Closes all streams and refuses new connections. Can be re-opened using connect()
+
 
 ### `connection:new_stream()` <!-- --> {#connection:new_stream}
 
-Create a new [stream](#stream) on the connection.
+Create a new [stream](#stream) on the connection. Use new_stream to initiate a new http request.
```

Comment:
It can be for more than just creating a new request: for HTTP2 you can also create a new stream for use in stream prioritisation, or for a push promise. (though usually for a promise you would usually use the higher level api)